Mathlib/Data/Fintype/Pi.lean

```diff
@@ -163,3 +163,21 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
     (Finset.univ.pi fun a : α => (Finset.univ : Finset (β a))) = Finset.univ := by
   ext; simp
 #align finset.univ_pi_univ Finset.univ_pi_univ
+
+/-! ### Diagonal -/
+
+namespace Finset
+variable {ι : Type*} [DecidableEq (ι → α)] {s : Finset α} {f : ι → α}
+
+/-- The diagonal of a finset `s : Finset α` as a finset of functions `ι → α`. -/
+def piDiag (s : Finset α) (ι : Type*) [DecidableEq (ι → α)] : Finset (ι → α) := s.image (const ι)
+
+@[simp] lemma mem_piDiag : f ∈ s.piDiag ι ↔ ∃ a ∈ s, const ι a = f := mem_image
+
+lemma piDiag_subset_piFinset [DecidableEq ι] [Fintype ι] :
```

Comment:
Let's also add
```
lemma piFinset_filter_const [DecidableEq ι] [Fintype ι] :
    (Fintype.piFinset fun _ => s).filter (∃ a ∈ s, const ι a = ·) = s.piDiag ι := by
  aesop
```
since it's easy to be more specific about this inclusion.

You also get a different proof of the inclusion, but that's up to you.
```
lemma piDiag_subset_piFinset [DecidableEq ι] [Fintype ι] :
    s.piDiag ι ⊆ Fintype.piFinset fun _ ↦ s := by
  simp [←piFinset_filter_const]
```